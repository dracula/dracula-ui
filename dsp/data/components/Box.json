{
  "class": "component",
  "type": "page",
  "id": "Box",
  "name": "Box",
  "last_updated": "2021-03-14T00:08:30.329Z",
  "last_updated_by": "System",
  "description": "Box is the most primitive component of Dracula UI.\nUsing Box allows for consumers of the components library to compose\nmore complex patterns, components, and UIs.\n\nBox includes built-in Color and Spacing properties that make building\ncomplex components convenient and consistent.",
  "related_entity_ids": [],
  "tags": ["component"],
  "snippets": {
    "trigger": "box",
    "languages": {
      "javascriptreact": "<Box color=\"purple\">Dracula</Box>;\n",
      "html": "<div color=\"purple\" class=\"drac-box drac-bg-purple\">Dracula</div>\n"
    }
  },
  "ext_com_draculaui_variations": [
    {
      "title": "Usage",
      "html": "<div color=\"purple\" class=\"drac-box drac-bg-purple\">Dracula</div>\n",
      "react": "<Box color=\"purple\">Dracula</Box>;\n",
      "docs": "",
      "screenshot": "./dsp/assets/Box.png",
      "svgData": "./dsp/assets/svgs/Box.svg"
    },
    {
      "title": "Colors",
      "html": "<div color=\"pinkPurple\" class=\"drac-box drac-bg-pink-purple\">Pink Purple</div>\n",
      "react": "<Box color=\"pinkPurple\">Pink Purple</Box>;\n",
      "docs": "",
      "screenshot": "./dsp/assets/BoxColors.png",
      "svgData": "./dsp/assets/svgs/BoxColors.svg"
    },
    {
      "title": "Rounded",
      "html": "<div style=\"display: flex; flex-direction: row\">\n  <div\n    color=\"purple\"\n    rounded=\"none\"\n    style=\"width: 250px; height: 250px; margin-right: 20px\"\n    class=\"drac-box drac-bg-purple drac-rounded-none\"\n  ></div>\n  <div\n    color=\"purple\"\n    rounded=\"sm\"\n    style=\"width: 250px; height: 250px; margin-right: 20px\"\n    class=\"drac-box drac-bg-purple drac-rounded-sm\"\n  ></div>\n  <div\n    color=\"purple\"\n    rounded=\"base\"\n    style=\"width: 250px; height: 250px; margin-right: 20px\"\n    class=\"drac-box drac-bg-purple drac-rounded\"\n  ></div>\n  <div\n    color=\"purple\"\n    rounded=\"lg\"\n    style=\"width: 250px; height: 250px; margin-right: 20px\"\n    class=\"drac-box drac-bg-purple drac-rounded-lg\"\n  ></div>\n  <div\n    color=\"purple\"\n    rounded=\"xl\"\n    style=\"width: 250px; height: 250px; margin-right: 20px\"\n    class=\"drac-box drac-bg-purple drac-rounded-xl\"\n  ></div>\n  <div\n    color=\"purple\"\n    rounded=\"2xl\"\n    style=\"width: 250px; height: 250px; margin-right: 20px\"\n    class=\"drac-box drac-bg-purple drac-rounded-2xl\"\n  ></div>\n  <div\n    color=\"purple\"\n    rounded=\"3xl\"\n    style=\"width: 250px; height: 250px; margin-right: 20px\"\n    class=\"drac-box drac-bg-purple drac-rounded-3xl\"\n  ></div>\n  <div\n    color=\"purple\"\n    rounded=\"full\"\n    style=\"width: 250px; height: 250px; margin-right: 20px\"\n    class=\"drac-box drac-bg-purple drac-rounded-full\"\n  ></div>\n</div>\n",
      "react": "<div style={{ display: \"flex\", flexDirection: \"row\" }}>\n  <Box\n    color=\"purple\"\n    rounded=\"none\"\n    style={{ width: 250, height: 250, marginRight: 20 }}\n  />\n  <Box\n    color=\"purple\"\n    rounded=\"sm\"\n    style={{ width: 250, height: 250, marginRight: 20 }}\n  />\n  <Box\n    color=\"purple\"\n    rounded=\"base\"\n    style={{ width: 250, height: 250, marginRight: 20 }}\n  />\n  <Box\n    color=\"purple\"\n    rounded=\"lg\"\n    style={{ width: 250, height: 250, marginRight: 20 }}\n  />\n  <Box\n    color=\"purple\"\n    rounded=\"xl\"\n    style={{ width: 250, height: 250, marginRight: 20 }}\n  />\n  <Box\n    color=\"purple\"\n    rounded=\"2xl\"\n    style={{ width: 250, height: 250, marginRight: 20 }}\n  />\n  <Box\n    color=\"purple\"\n    rounded=\"3xl\"\n    style={{ width: 250, height: 250, marginRight: 20 }}\n  />\n  <Box\n    color=\"purple\"\n    rounded=\"full\"\n    style={{ width: 250, height: 250, marginRight: 20 }}\n  />\n</div>;\n",
      "docs": "",
      "screenshot": "./dsp/assets/BoxRounded.png",
      "svgData": "./dsp/assets/svgs/BoxRounded.svg"
    },
    {
      "title": "Generic Elements",
      "html": "<div>\n  <button as=\"button\" color=\"pinkPurple\" class=\"drac-box drac-bg-pink-purple\">\n    Button</button\n  ><a as=\"a\" color=\"green\" href=\"#\" class=\"drac-box drac-bg-green\">Link</a>\n</div>\n",
      "react": "<div>\n  <Box as=\"button\" color=\"pinkPurple\">\n    Button\n  </Box>\n  <Box as=\"a\" color=\"green\" href=\"#\">\n    Link\n  </Box>\n</div>;\n",
      "docs": "",
      "screenshot": "./dsp/assets/BoxGeneric Elements.png",
      "svgData": "./dsp/assets/svgs/BoxGeneric Elements.svg"
    }
  ],
  "ext_com_draculaui_props": {
    "color": {
      "defaultValue": null,
      "description": "The background color.",
      "name": "color",
      "parent": {
        "fileName": "src/components/Box/Box.tsx",
        "name": "BoxProps"
      },
      "declarations": [
        { "fileName": "src/components/Box/Box.tsx", "name": "BoxProps" }
      ],
      "required": false,
      "type": {
        "name": "\"white\" | \"cyan\" | \"green\" | \"orange\" | \"pink\" | \"purple\" | \"red\" | \"yellow\" | \"purpleCyan\" | \"yellowPink\" | \"cyanGreen\" | \"pinkPurple\" | \"animated\" | undefined"
      }
    },
    "rounded": {
      "defaultValue": null,
      "description": "The border radius.",
      "name": "rounded",
      "parent": {
        "fileName": "src/components/Box/Box.tsx",
        "name": "BoxProps"
      },
      "declarations": [
        { "fileName": "src/components/Box/Box.tsx", "name": "BoxProps" }
      ],
      "required": false,
      "type": {
        "name": "\"none\" | \"sm\" | \"base\" | \"lg\" | \"xl\" | \"2xl\" | \"3xl\" | \"full\" | undefined"
      }
    },
    "spacing": {
      "defaultValue": null,
      "description": "Dracula UI standard spacing properties.",
      "name": "spacing",
      "parent": {
        "fileName": "src/components/Box/Box.tsx",
        "name": "BoxProps"
      },
      "declarations": [
        { "fileName": "src/components/Box/Box.tsx", "name": "BoxProps" }
      ],
      "required": false,
      "type": { "name": "SpacingPropType | undefined" }
    },
    "margin": {
      "defaultValue": null,
      "description": "Dracula UI standard margin properties.",
      "name": "margin",
      "parent": {
        "fileName": "src/components/Box/Box.tsx",
        "name": "BoxProps"
      },
      "declarations": [
        { "fileName": "src/components/Box/Box.tsx", "name": "BoxProps" }
      ],
      "required": false,
      "type": { "name": "MarginPropType | undefined" }
    },
    "as": {
      "defaultValue": null,
      "description": "The HTML element to be used",
      "name": "as",
      "parent": {
        "fileName": "src/components/Box/Box.tsx",
        "name": "BoxProps"
      },
      "declarations": [
        { "fileName": "src/components/Box/Box.tsx", "name": "BoxProps" }
      ],
      "required": false,
      "type": { "name": "keyof HTMLElementTagNameMap | undefined" }
    }
  },
  "ext_com_draculaui_docgen": {
    "tags": {},
    "description": "Box is the most primitive component of Dracula UI.\nUsing Box allows for consumers of the components library to compose\nmore complex patterns, components, and UIs.\n\nBox includes built-in Color and Spacing properties that make building\ncomplex components convenient and consistent.",
    "displayName": "Box",
    "methods": [],
    "props": {
      "color": {
        "defaultValue": null,
        "description": "The background color.",
        "name": "color",
        "parent": {
          "fileName": "src/components/Box/Box.tsx",
          "name": "BoxProps"
        },
        "declarations": [
          { "fileName": "src/components/Box/Box.tsx", "name": "BoxProps" }
        ],
        "required": false,
        "type": {
          "name": "\"white\" | \"cyan\" | \"green\" | \"orange\" | \"pink\" | \"purple\" | \"red\" | \"yellow\" | \"purpleCyan\" | \"yellowPink\" | \"cyanGreen\" | \"pinkPurple\" | \"animated\" | undefined"
        }
      },
      "rounded": {
        "defaultValue": null,
        "description": "The border radius.",
        "name": "rounded",
        "parent": {
          "fileName": "src/components/Box/Box.tsx",
          "name": "BoxProps"
        },
        "declarations": [
          { "fileName": "src/components/Box/Box.tsx", "name": "BoxProps" }
        ],
        "required": false,
        "type": {
          "name": "\"none\" | \"sm\" | \"base\" | \"lg\" | \"xl\" | \"2xl\" | \"3xl\" | \"full\" | undefined"
        }
      },
      "spacing": {
        "defaultValue": null,
        "description": "Dracula UI standard spacing properties.",
        "name": "spacing",
        "parent": {
          "fileName": "src/components/Box/Box.tsx",
          "name": "BoxProps"
        },
        "declarations": [
          { "fileName": "src/components/Box/Box.tsx", "name": "BoxProps" }
        ],
        "required": false,
        "type": { "name": "SpacingPropType | undefined" }
      },
      "margin": {
        "defaultValue": null,
        "description": "Dracula UI standard margin properties.",
        "name": "margin",
        "parent": {
          "fileName": "src/components/Box/Box.tsx",
          "name": "BoxProps"
        },
        "declarations": [
          { "fileName": "src/components/Box/Box.tsx", "name": "BoxProps" }
        ],
        "required": false,
        "type": { "name": "MarginPropType | undefined" }
      },
      "as": {
        "defaultValue": null,
        "description": "The HTML element to be used",
        "name": "as",
        "parent": {
          "fileName": "src/components/Box/Box.tsx",
          "name": "BoxProps"
        },
        "declarations": [
          { "fileName": "src/components/Box/Box.tsx", "name": "BoxProps" }
        ],
        "required": false,
        "type": { "name": "keyof HTMLElementTagNameMap | undefined" }
      }
    }
  }
}
