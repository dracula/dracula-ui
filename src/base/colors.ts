import mapValues from 'lodash/mapValues'

export const supportColors = {
  white: 'drac-bg-white',
  black: 'drac-bg-black',
  blackSecondary: 'drac-bg-black-secondary',
  blackLight: 'drac-bg-black-light',
  grey: 'drac-bg-grey',
  greySecondary: 'drac-bg-grey-secondary',
  greyLight: 'drac-bg-grey-light'
}

export const baseColors = {
  cyan: 'drac-bg-cyan',
  green: 'drac-bg-green',
  orange: 'drac-bg-orange',
  pink: 'drac-bg-pink',
  purple: 'drac-bg-purple',
  red: 'drac-bg-red',
  yellow: 'drac-bg-yellow'
}

export const gradientColors = {
  purpleCyan: 'drac-bg-purple-cyan',
  yellowPink: 'drac-bg-yellow-pink',
  cyanGreen: 'drac-bg-cyan-green',
  pinkPurple: 'drac-bg-pink-purple'
}

export const animatedColors = {
  animated: 'drac-bg-animated'
}

export const colors = {
  ...supportColors,
  ...baseColors,
  ...gradientColors
}

export const backgroundColors = {
  ...colors,
  ...animatedColors
}

export type ColorNames = keyof typeof colors
export type ColorMap = Record<ColorNames, string>

export type SupportColorNames = keyof typeof supportColors
export type SupportColorMap = Record<SupportColorNames, string>

export type BaseColorNames = keyof typeof baseColors
export type BaseColorMap = Record<BaseColorNames, string>

export type GradientColorNames = keyof typeof gradientColors
export type GradientBaseColorMap = Record<GradientColorNames, string>

export const borderColors: BaseColorMap = mapValues(baseColors, (className) =>
  className.replace('-bg-', '-border-')
)

export const glowColors: BaseColorMap = mapValues(baseColors, (className) =>
  className.replace('-bg-', '-glow-')
)

export const baseTextColors: ColorMap = mapValues(colors, (className) => {
  return className.replace('-bg-', '-text-')
})

export const scrollbarColors: ColorMap = mapValues(
  colors,
  (className) => {
    return className.replace('-bg-', '-scrollbar-')
  }
)

export const colorUtilities = {
  classes: [
    ...Object.values(baseColors),
    ...Object.values(baseTextColors),
    ...Object.values(supportColors),
    ...Object.values(gradientColors),
    ...Object.values(borderColors),
    ...Object.values(animatedColors)
  ],
  react: {
    base: Object.keys(baseColors),
    text: Object.keys(baseTextColors),
    support: Object.keys(supportColors),
    gradient: Object.keys(gradientColors),
    border: Object.keys(borderColors),
    animated: Object.keys(animatedColors)
  }
}
