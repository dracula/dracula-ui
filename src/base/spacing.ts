import { mapValues, omit } from 'lodash'

const padding = {
  none: 'drac-p-none',
  xxs: 'drac-p-xxs',
  xs: 'drac-p-xs',
  sm: 'drac-p-sm',
  md: 'drac-p-md',
  lg: 'drac-p-lg'
}

const paddingLeft = mapValues(padding, (clz) => clz.replace('-p-', '-pl-'))
const paddingRight = mapValues(padding, (clz) => clz.replace('-p-', '-pr-'))
const paddingTop = mapValues(padding, (clz) => clz.replace('-p-', '-pt-'))
const paddingBottom = mapValues(padding, (clz) => clz.replace('-p-', '-pb-'))
const paddingX = mapValues(padding, (clz) => clz.replace('-p-', '-px-'))
const paddingY = mapValues(padding, (clz) => clz.replace('-p-', '-py-'))

type paddingType = keyof typeof padding
export type PaddingMixin = {
  p?: paddingType

  py?: paddingType
  px?: paddingType

  pt?: paddingType
  pb?: paddingType

  pl?: paddingType
  pr?: paddingType
}

export function paddingMixin(mixin: PaddingMixin) {
  let classes: string[] = []

  if (mixin.p) {
    classes.push(padding[mixin.p])
  }

  if (mixin.py) {
    classes.push(paddingY[mixin.py])
  }

  if (mixin.px) {
    classes.push(paddingX[mixin.px])
  }

  if (mixin.pt) {
    classes.push(paddingTop[mixin.pt])
  }

  if (mixin.pb) {
    classes.push(paddingBottom[mixin.pb])
  }

  if (mixin.pl) {
    classes.push(paddingLeft[mixin.pl])
  }

  if (mixin.pr) {
    classes.push(paddingRight[mixin.pr])
  }

  return classes
}

const margin = {
  none: 'drac-m-none',
  xxs: 'drac-m-xxs',
  xs: 'drac-m-xs',
  sm: 'drac-m-sm',
  md: 'drac-m-md',
  lg: 'drac-m-lg'
}

const mLeft = mapValues(margin, (clz) => clz.replace('-m-', '-ml-'))
const mRight = mapValues(margin, (clz) => clz.replace('-m-', '-mr-'))
const mTop = mapValues(margin, (clz) => clz.replace('-m-', '-mt-'))
const mBottom = mapValues(margin, (clz) => clz.replace('-m-', '-mb-'))
const mX = mapValues(margin, (clz) => clz.replace('-m-', '-mx-'))
const mY = mapValues(margin, (clz) => clz.replace('-m-', '-my-'))

type marginType = keyof typeof margin
export type MarginMixin = {
  m?: marginType

  my?: marginType
  mx?: marginType

  mt?: marginType
  mb?: marginType

  ml?: marginType
  mr?: marginType
}

export function marginMixin(mixin: MarginMixin) {
  let classes: string[] = []

  if (mixin.m) {
    classes.push(margin[mixin.m])
  }

  if (mixin.my) {
    classes.push(mY[mixin.my])
  }

  if (mixin.mx) {
    classes.push(mX[mixin.mx])
  }

  if (mixin.mt) {
    classes.push(mTop[mixin.mt])
  }

  if (mixin.mb) {
    classes.push(mBottom[mixin.mb])
  }

  if (mixin.ml) {
    classes.push(mLeft[mixin.ml])
  }

  if (mixin.mr) {
    classes.push(mRight[mixin.mr])
  }

  return classes
}

export const spacingUtilities = {
  classes: {
    padding: [
      ...Object.values(padding),
      ...Object.values(paddingX),
      ...Object.values(paddingY),
      ...Object.values(paddingTop),
      ...Object.values(paddingBottom)
    ],
    margin: [
      ...Object.values(margin),
      ...Object.values(mX),
      ...Object.values(mY),
      ...Object.values(mTop),
      ...Object.values(mBottom)
    ]
  },
  react: {
    padding: [`p`, `px`, `py`, `pt`, `pb`, `pl`, `pr`],
    margin: [`m`, `mx`, `my`, `mt`, `mb`, `ml`, `mr`]
  }
}

/**
 * Removes all spacing props from props object
 */
export function cleanProps<T>(
  props: T & Partial<PaddingMixin> & Partial<MarginMixin>
) {
  return omit(props, [
    ...spacingUtilities.react.padding,
    ...spacingUtilities.react.margin
  ]) as Omit<T, keyof PaddingMixin | keyof MarginMixin>
}
