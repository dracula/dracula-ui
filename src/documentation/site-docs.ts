import {
  ComponentExample,
  renderSnapshot,
  SnapshotBuilder
} from './render-component'
import React from 'react'
import fs from 'fs-extra'
import { pretty } from './pretty'
import { getDocGen } from '../documentation/docgen/doc-generator'
import { toDSP } from '../documentation/dsp/component-generator'
import puppeteer from 'puppeteer'
import { componentScreenshot } from './screenshots/component-screenshot'
import os from 'os'

interface Documentation {
  basic: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder> | SnapshotBuilder
}

jest.setTimeout(10000)

const RECORDING = process.env.RECORDING

export function docs<T>(
  componentClass: React.FunctionComponent<T> | string,
  documentation: Documentation
) {
  const name =
    typeof componentClass === 'string'
      ? componentClass
      : componentClass.displayName ?? componentClass.name

  const dspPath = `${process.cwd()}/dsp`

  describe(`Site: ${name}`, () => {
    let browser: puppeteer.Browser

    beforeAll(async () => {
      if (!RECORDING) {
        return
      }

      let executablePath = undefined
      if (os.arch() === 'arm64') {
        executablePath = '/Applications/Chromium.app/Contents/MacOS/Chromium'
      }

      browser = await puppeteer.launch({
        headless: true,
        devtools: true,
        executablePath
      })
    })

    let examples: Record<string, ComponentExample> = {}

    afterAll(async () => {
      if (!RECORDING) {
        return
      }

      await browser.close()
      const docGen = getDocGen(name)
      const dsp = toDSP(name, examples, docGen)

      await fs.mkdirp(`${dspPath}/data/components/`).catch(() => {
        // folder already exists
      })

      const path = `${dspPath}/data/components/${name}.json`
      await fs.writeFile(path, pretty(JSON.stringify(dsp), 'json'))
    })

    test('Basic Usage', async () => {
      const variation = documentation.basic()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp,
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()

      if (RECORDING) {
        const [screenshot, svg] = await componentScreenshot(
          browser,
          snapshot,
          null,
          name
        )

        snapshot.screenshot = screenshot
        snapshot.svgData = svg
      }

      examples = {
        ...examples,
        basic: snapshot
      }
    })

    const variations: SnapshotBuilder[] = ([] as SnapshotBuilder[]).concat(
      documentation.variations()
    )

    variations.forEach((variation) => {
      test(variation.title, async () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp,
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()

        if (RECORDING) {
          const [screenshot, svg] = await componentScreenshot(
            browser,
            snapshot,
            variation,
            name
          )

          snapshot.screenshot = screenshot
          snapshot.svgData = svg
        }

        examples = {
          ...examples,
          [variation.title]: snapshot
        }
      })
    })
  })
}
