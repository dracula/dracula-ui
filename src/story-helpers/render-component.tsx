import React from 'react'
import Enzyme, { mount, shallow } from 'enzyme'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import { pretty } from './pretty'

Enzyme.configure({ adapter: new Adapter() })

export type ComponentExample = {
  html: string
  docs: string
  title: string
  react?: any
  screenshot?: string
  svgData?: string
}

export type SnapshotBuilder = {
  comp: () => React.ReactElement
  docs: string
  title: string
}

export function buildSnapshot(
  title: string,
  comp: () => React.ReactElement,
  docs: string = ''
): SnapshotBuilder {
  return {
    title,
    comp,
    docs: docs.trim()
  }
}

export function renderSnapshot(
  title: string,
  comp: () => React.ReactElement,
  docs: string = ''
): ComponentExample {
  let component = comp()
  const html = mount(component).html()
  let wrapped = false

  // wrap the component in a div when it's not wrapped already
  // we do this so we can get the actual component example to not be shallow rendered.
  // Shallow rendering a parent "div" gives us a nice JSX representation of the example.
  if (component.type !== 'div') {
    component = <div>{component}</div>
    wrapped = true
  }

  let react = shallow(component).debug({ verbose: true })

  if (wrapped) {
    // replace wrapper div ðŸ˜¬
    react = react
      .replace(/^<div>\n/, '')
      .replace(/\n<\/div>$/, '')
      .trim()
  }

  return {
    title,
    html: pretty(html, 'html'),
    react: pretty(react, 'typescript'),
    docs
  }
}
