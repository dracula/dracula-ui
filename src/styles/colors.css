:root {
  --black: hsl(230, 15%, 15%);
  --blackSecondary: hsl(230, 15%, 30%);
  --blackTernary: hsl(230, 15%, 70%);
  --blackLight: hsla(230, 15%, 15%, 5%);

  --grey: hsl(230, 15%, 30%);
  --greySecondary: hsl(230, 15%, 45%);
  --greyTernary: hsl(230, 15%, 75%);
  --greyLight: hsla(230, 15%, 30%, 5%);

  --white: hsl(60, 30%, 96%);
  --whiteSecondary: hsl(60, 30%, 100%);
  --whiteLight: hsla(60, 30%, 96%, 5%);

  --cyan: hsl(170, 100%, 75%);
  --cyanSecondary: hsl(170, 100%, 90%);
  --cyanLight: hsla(170, 100%, 75%, 5%);

  --green: hsl(115, 100%, 75%);
  --greenSecondary: hsl(115, 100%, 90%);
  --greenLight: hsla(115, 100%, 75%, 5%);

  --orange: hsl(35, 100%, 75%);
  --orangeSecondary: hsl(35, 100%, 90%);
  --orangeLight: hsla(35, 100%, 75%, 5%);

  --pink: hsl(330, 100%, 75%);
  --pinkSecondary: hsl(330, 100%, 90%);
  --pinkLight: hsla(330, 100%, 75%, 5%);

  --purple: hsl(250, 100%, 75%);
  --purpleSecondary: hsl(250, 100%, 90%);
  --purpleLight: hsl(250, 100%, 75%, 5%);

  --red: hsl(10, 100%, 75%);
  --redSecondary: hsl(10, 100%, 90%);
  --redLight: hsla(10, 100%, 75%, 5%);

  --yellow: hsl(60, 100%, 75%);
  --yellowSecondary: hsl(60, 100%, 90%);
  --yellowLight: hsla(60, 100%, 75%, 5%);

  --cyan-transparent: hsl(170, 100%, 75%, 0.1);
  --green-transparent: hsl(115, 100%, 75%, 0.1);
  --orange-transparent: hsl(35, 100%, 75%, 0.1);
  --pink-transparent: hsl(330, 100%, 75%, 0.1);
  --purple-transparent: hsl(250, 100%, 75%, 0.1);
  --red-transparent: hsl(10, 100%, 75%, 0.1);
  --yellow-transparent: hsl(60, 100%, 75%, 0.1);

  --disabled: hsl(230, 15%, 50%);
  --gradientDegree: 135deg;

  --glowColor: rgba(255, 255, 255, 0.25);

  @each $opacity in 10, 20, 30, 40, 50, 60, 70, 80, 90 {
    --cyan-$(opacity)0: hsl(170, 100%, 75%, 0.$(opacity));
    --green-$(opacity)0: hsl(115, 100%, 75%, 0.$(opacity));
    --orange-$(opacity)0: hsl(35, 100%, 75%, 0.$(opacity));
    --pink-$(opacity)0: hsl(330, 100%, 75%, 0.$(opacity));
    --purple-$(opacity)0: hsl(250, 100%, 75%, 0.$(opacity));
    --red-$(opacity)0: hsl(10, 100%, 75%, 0.$(opacity));
    --yellow-$(opacity)0: hsl(60, 100%, 75%, 0.$(opacity));
  }

  --accentColor: var(--purple);

  @each $from, $to in (purple, yellow, cyan, pink), (cyan, pink, green, purple)
  {
    --$(from)-$(to): linear-gradient(
      var(--gradientDegree),
      var(--$(from)) 0%,
      var(--$(to)) 100%
    );
  }
}

.bg-animated {
  content: '';
  animation: animatedGradient 6s ease infinite alternate;
  background-size: 300% 300%;
  background-image: linear-gradient(
    var(--gradientDegree),
    var(--orange),
    var(--yellow),
    var(--red),
    var(--pink),
    var(--purple),
    var(--cyan),
    var(--green)
  );
}

@keyframes animatedGradient {
  0% {
    background-position: 0 50%;
  }
  50% {
    background-position: 100% 50%;
  }
  to {
    background-position: 0 50%;
  }
}

