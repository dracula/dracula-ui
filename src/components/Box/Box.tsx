import cx from 'classnames/dedupe'
import React, { AllHTMLAttributes } from 'react'
import { borderColors, colors, glowColors } from '../../base/colors'
import {
  PaddingMixin,
  paddingMixin,
  marginMixin,
  MarginMixin
} from '../../base/spacing'

type Element = HTMLElementTagNameMap

export const roundedBorders = {
  none: 'drac-rounded-none',
  sm: 'drac-rounded-sm',
  base: 'drac-rounded',
  lg: 'drac-rounded-lg',
  xl: 'drac-rounded-xl',
  '2xl': 'drac-rounded-2xl',
  '3xl': 'drac-rounded-3xl',
  full: 'drac-rounded-full'
}

/**
 * Box Props
 */
export type BoxProps<K extends keyof Element = 'div'> = {
  /** The background color. */
  color?: keyof typeof colors

  /** The glow color. */
  glowColor?: keyof typeof glowColors

  /** The border color. */
  borderColor?: keyof typeof borderColors

  /** The border radius. */
  rounded?: keyof typeof roundedBorders

  /** The HTML element to be used */
  as?: K
} & AllHTMLAttributes<K> &
  PaddingMixin &
  MarginMixin

/**
 * Box is the most primitive component of Dracula UI.
 * Using Box allows for consumers of the components library to compose
 * more complex patterns, components, and UIs.
 *
 * Box includes built-in Color and Spacing properties that make building
 * complex components convenient and consistent.
 */
export function Box<T extends keyof Element>(props: BoxProps<T>) {
  const finalProps = {
    ...props,
    className: cx(
      `drac-box`,
      props.className,
      props.color && colors[props.color],
      props.glowColor && glowColors[props.glowColor],
      props.borderColor && borderColors[props.borderColor],
      props.rounded && roundedBorders[props.rounded],
      ...paddingMixin(props),
      ...marginMixin(props)
    )
  }

  const as = finalProps.as ?? 'div'
  return React.createElement(as, finalProps, props.children)
}

Box.displayName = 'Box'
