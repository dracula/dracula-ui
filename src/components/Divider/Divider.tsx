import { borderColors } from '../../base/colors'
import React, { HTMLAttributes } from 'react'

export const dividerColors = borderColors

/**
 * Divider Props
 */
export interface DividerProps extends HTMLAttributes<HTMLDivElement> {
  /** The Dracula UI color for the Divider. */
  color: keyof typeof dividerColors
}

/**
 * Dividers are horizontal lines used to separate semantic blocks of
 * content or UI patterns.
 */
export const Divider: React.FC<DividerProps> = (props: DividerProps) => {
  const finalProps = {
    ...props,
    className: `drac-divider ${dividerColors[props.color]}`
  }

  return <hr {...finalProps} />
}

Divider.displayName = 'Divider'
