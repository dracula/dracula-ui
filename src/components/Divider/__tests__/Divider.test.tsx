import React from 'react'
import { docs } from '@/documentation/site-docs'
import { snapshot } from '@/documentation/render-component'
import { Box } from '@/components/Box/Box'
import { baseColors } from '@/base/colors'
import { Divider, DividerProps } from '@/components/Divider/Divider'

docs(Divider, {
  basic() {
    return snapshot(
      'Usage',
      Usage,
      'Dividers are used to define a thematic break in a page.'
    )
  },
  variations() {
    return [
      snapshot(
        'Colors',
        Colors,
        'Dividers can be customized to use any of the Dracula UI theme colors.'
      )
    ]
  }
})

function Usage() {
  return (
    <Divider />
  )
}

function Colors() {
  return (
    <Box>
      {Object.keys(baseColors).map((color) => {
        return (
          <Box key={color} mb="sm">
            <Divider color={color as DividerProps['color']} />
          </Box>
        )
      })}
    </Box>
  )
}