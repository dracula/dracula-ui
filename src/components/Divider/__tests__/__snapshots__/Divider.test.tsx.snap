// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site: Divider Basic Usage 1`] = `
Object {
  "docs": "Dividers are used to define a thematic break in a page.",
  "html": "<hr class=\\"drac-divider\\" />
",
  "react": "<Divider />;
",
  "title": "Usage",
}
`;

exports[`Site: Divider Colors 1`] = `
Object {
  "docs": "Dividers can be customized to use any of the Dracula UI theme colors.",
  "html": "<div class=\\"drac-box\\">
  <div class=\\"drac-box drac-mb-sm\\">
    <hr class=\\"drac-divider drac-border-cyan\\" />
  </div>
  <div class=\\"drac-box drac-mb-sm\\">
    <hr class=\\"drac-divider drac-border-green\\" />
  </div>
  <div class=\\"drac-box drac-mb-sm\\">
    <hr class=\\"drac-divider drac-border-orange\\" />
  </div>
  <div class=\\"drac-box drac-mb-sm\\">
    <hr class=\\"drac-divider drac-border-pink\\" />
  </div>
  <div class=\\"drac-box drac-mb-sm\\">
    <hr class=\\"drac-divider drac-border-purple\\" />
  </div>
  <div class=\\"drac-box drac-mb-sm\\">
    <hr class=\\"drac-divider drac-border-red\\" />
  </div>
  <div class=\\"drac-box drac-mb-sm\\">
    <hr class=\\"drac-divider drac-border-yellow\\" />
  </div>
</div>
",
  "react": "<Box>
  <Box mb=\\"sm\\">
    <Divider color=\\"cyan\\" />
  </Box>
  <Box mb=\\"sm\\">
    <Divider color=\\"green\\" />
  </Box>
  <Box mb=\\"sm\\">
    <Divider color=\\"orange\\" />
  </Box>
  <Box mb=\\"sm\\">
    <Divider color=\\"pink\\" />
  </Box>
  <Box mb=\\"sm\\">
    <Divider color=\\"purple\\" />
  </Box>
  <Box mb=\\"sm\\">
    <Divider color=\\"red\\" />
  </Box>
  <Box mb=\\"sm\\">
    <Divider color=\\"yellow\\" />
  </Box>
</Box>;
",
  "title": "Colors",
}
`;
